src/linear_algebra/dual.lean

```diff
@@ -398,3 +407,145 @@ begin
 end
 
 end dual_pair
+
+namespace submodule
+
+universes u v w
+
+variables {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M] [module R M]
+variable {W : submodule R M}
+
+/-- The `dual_restrict` of a submodule `W` of `M` is the linear map from the
+  dual of `M` to the dual of `W` such that the domain of each linear map is
+  restricted to `W`. -/
+def dual_restrict (W : submodule R M) :
+  module.dual R M →ₗ[R] module.dual R W :=
+linear_map.dom_restrict' W
+
+@[simp] lemma dual_restrict_apply
+  (W : submodule R M) (φ : module.dual R M) (x : W) :
+  W.dual_restrict φ x = φ x.1 := rfl
```

Comment:
To go from a submodule to the whole space (generally: from a subtype to the whole type), mathlib uses coercions rather than projections. So like this we can avoid an extra `simp` step:
```suggestion
  W.dual_restrict φ x = φ (x : M) := rfl
```